src/libraries/System.Net.WebSockets/src/System/Net/WebSockets/ManagedWebSocket.SendLock.cs

```diff
@@ -0,0 +1,156 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+using System.Threading;
+using System.Threading.Tasks;
+using System.Threading.Tasks.Sources;
+
+namespace System.Net.WebSockets
+{
+    internal partial class ManagedWebSocket
+    {
+        /// <summary>
+        /// Lightweight async lock that allows single owner at any given time.
+        /// </summary>
+        [StructLayout(LayoutKind.Auto)]
+        private struct SendLock
```

Comment:
I am concerned with the addition of this custom lock... are you sure there aren't any existing locks you could use?

Reply:
This is equivalent of SemaphoreSlim(1, 1), but optimized for the case where there is no contention for the lock. I did extensive profiling and the semaphore that was used before used about 10-15% of the CPU in some of the cases.

Because the WebSocket explicitly doesn't support concurrent SendAsync, the only time where there might be contending for this lock is the case when the socket has received a control message and needs to reply and in the mean time there might be in flight send operation.

I am not aware of any lock primitive other than SemaphoreSlim, that allows acquiring in one thread and releasing in another.

Reply:
cc @stephentoub